### PR TITLE
Case 21876: New procedural uniforms for time, iEntityTime and iLocalCreatedTime

### PR DESCRIPTION
**Pull request by SamGondelman**
https://highfidelity.manuscript.com/f/cases/21876/Expose-iEntityTime-and-iLocalCreatedTime-for-procedurals

Test plan:
- Go to a domain and run this:
```
var entity = Entities.addEntity({
	type: "Sphere",
             position: Vec3.sum(MyAvatar.position, Vec3.multiply(2, Quat.getFront(MyAvatar.orientation))),
             dimensions: 0.5,
             collisionless: true,
             userData: JSON.stringify({
                  "ProceduralEntity": {
                       "version": 2,
                       "shaderUrl": "https://hifi-content.s3.amazonaws.com/samuel/ProceduralTime.fs”
                  }
             })
});
```
- You'll see a sphere split into 3 sections: red, green, and blue.  They should all be pulsing, but could be slightly off from each other.
- Red corresponds to iGlobalTime, which is the time since the shader last compiled.  If you change the shaderURL or uniforms, etc., the shader will be recompiled, and the red will restart.
- Green corresponds to iLocalCreatedTime, the time since the first shader for this object was compiled.  If you do something to cause the shader to recompile, this will not restart.
- Blue corresponds to iEntityTime, the time since the entity creation.  This is shared across clients and can't change, so you can create shader effects that are synced across clients.
- Have someone else running this PR join you.  They'll see the sphere too, but the timing of the pulses will be slightly off.  The blue pulses should pretty closely match.
- Run this:
```
Entities.editEntity(entity, {
     userData: JSON.stringify({
          "ProceduralEntity": {
               "version": 2,
               "uniforms": {
                    "_shininess": 1
               },
               "shaderUrl": "https://hifi-content.s3.amazonaws.com/samuel/ProceduralTime.fs?8"
          }
     })
});
```
Immediately after you run that, the red pulses should restart for both of you, but the other pulses shouldn't be affected.  This will only happen the first time you run that snippet, if you don't catch it, you can run the first snippet again and try again.

### CODE REVIEW

**Comment by hifi-gustavo** [comment]
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/18309/HighFidelity-Beta-PR15243-4fd2c44.dmg

**Comment by hifi-gustavo** [comment]
Mac OS X client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/18309/HighFidelity-Beta-Interface-PR15243-4fd2c44.dmg

**Comment by hifi-gustavo** [comment]
Android build is available [here](https://deployment.highfidelity.com/jobs/pr-build/label%3Dandroid/18309/HighFidelity-Beta-PR15243-4fd2c44.apk).  Quest build is available [here](https://deployment.highfidelity.com/jobs/pr-build/label%3Dandroid/18309/HighFidelity-Quest-Beta-PR15243-4fd2c44.apk)

**Comment by hifi-gustavo** [comment]
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/18309/HighFidelity-Beta-PR15243-4fd2c44.exe

**Comment by hifi-gustavo** [comment]
Windows client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/18309/HighFidelity-Beta-Interface-PR15243-4fd2c44.exe

File: libraries/procedural/src/procedural/ProceduralCommon.slh [hunk 1]
```diff
@@ -52,6 +54,8 @@ LAYOUT_STD140(binding=0) uniform standardInputsBuffer {
 #define iWorldPosition standardInputs.worldPosition
 #define iWorldScale standardInputs.worldScale
 #define iGlobalTime standardInputs.globalTime
+#define iLocalCreatedTime standardInputs.localCreatedTime
+#define iEntityTime standardInputs.entityTime
```
Comment (SamGondelman) [inline libraries/procedural/src/procedural/ProceduralCommon.slh]: I'm open to suggestions for alternate names for these.

File: libraries/procedural/src/procedural/Procedural.cpp [hunk 2]
```diff
@@ -225,11 +225,13 @@ void Procedural::prepare(gpu::Batch& batch,
                          const glm::vec3& position,
                          const glm::vec3& size,
                          const glm::quat& orientation,
+                         const quint64& created,
```
Comment (AndrewMeadows) [inline libraries/procedural/src/procedural/Procedural.cpp]: Not a blocker, but I prefer `uint64_t` over of `quint64`.  We should remove as many Qt-isms as possible from code that does not **need** to depend on Qt.

File: libraries/procedural/src/procedural/ProceduralCommon.slh (hunk 1 above)
Comment (AndrewMeadows) [reply libraries/procedural/src/procedural/ProceduralCommon.slh]: Are procedurals only ever to be applied to `EntityItem`s?  If so then maybe `iEntityTime` makes sense, if not: then we should avoid all mentions of "entity" in the prodedural code.

Comment (SamGondelman) [reply libraries/procedural/src/procedural/ProceduralCommon.slh]: The eventual goal is to have Material Entities be the interface by which you apply procedurals to things in the world.  So this value would always be relative to the material entity's `created`, regardless of if it's applied to an entity or an avatar.

**Comment by hifi-gustavo** [comment]
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/18367/HighFidelity-Beta-PR15243-de5643b.dmg

**Comment by hifi-gustavo** [comment]
Mac OS X client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/18367/HighFidelity-Beta-Interface-PR15243-de5643b.dmg

**Comment by hifi-gustavo** [comment]
Android build is available [here](https://deployment.highfidelity.com/jobs/pr-build/label%3Dandroid/18367/HighFidelity-Beta-PR15243-de5643b.apk).  Quest build is available [here](https://deployment.highfidelity.com/jobs/pr-build/label%3Dandroid/18367/HighFidelity-Quest-Beta-PR15243-de5643b.apk)

**Comment by hifi-gustavo** [comment]
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/18367/HighFidelity-Beta-PR15243-de5643b.exe

**Comment by hifi-gustavo** [comment]
Windows client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/18367/HighFidelity-Beta-Interface-PR15243-de5643b.exe

**Pull request merged**